src/core/Akka/Actor/Exceptions.cs

```diff
@@ -114,6 +114,40 @@ protected AskTimeoutException(SerializationInfo info, StreamingContext context)
 #endif
     }
 
+    /// <summary>
+    /// This exception is thrown when an Ask expects message of type A, but receives B.
+    /// </summary>
+    public class InvalidAskResponseException : AkkaException
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AskTimeoutException"/> class.
+        /// </summary>
+        /// <param name="message">The message that describes the error.</param>
+        /// <param name="responseMsg">The unexpected message object received by the actor.</param>
+        public InvalidAskResponseException(string message, object responseMsg)
+            : base(message)
+        {
+            ResponseMsg = responseMsg;
+        }
+
+        /// <summary>
+        /// The message received by the Ask
+        /// </summary>
+        public object ResponseMsg { get; }
```

Comment:
Error message includes a reference to the underlying "unexpected" message type, in case the end-user needs more information about why it failed.